File: C++/Minstack.cpp
#include<iostream>
#include<stack>

class Solution {
public:
    std::stack<int>stack1;
    std::stack<int>Min;
    void push(int value) {
        stack1.push(value);
        if(Min.empty() || value <= Min.top()){
            Min.push(value);
        }
    }
    
    void pop() {
        if (stack1.top() == Min.top()) {
            stack1.pop();
            Min.pop();
        } else {
            stack1.pop();
        }
    }
    
    int top() {
        return stack1.top();
    }
    
    int min() {
        return Min.top();
    }
};

int main(){
    Solution solution;
    solution.push(-1);
    solution.push(2);
    std::cout<<solution.min()<<" ";
    std::cout<<solution.top()<<" ";
    solution.pop();
    solution.push(1);
    std::cout<<solution.top()<<" ";
    std::cout<<solution.min()<<std::endl;

    return 0;
}

File: python/minStack.py
class Solution:
    def __init__(self):
        self.stack1=[]
        self.Min=[]
    def push(self, node):
        self.stack1.append(node)
        if not self.Min or node <= self.Min[-1]:
            self.Min.append(node)

    def pop(self):
        if self.stack1.pop() == self.Min[-1]:
            self.Min.pop()

    def top(self):
        return self.stack1[-1]

    def min(self):
        return self.Min[-1]

s=Solution()
s.push(-1)
s.push(2)
print(s.min())
print(s.top())
print(s.pop())
s.push(1)
print(s.top())
print(s.min())